Add: telnet:// URI scheme support for one-click MUD connections

MAVRICK-1 opened the pull request:
# Add telnet:// URI scheme support for one-click MUD connections

## Summary

This PR implements telnet:// URI scheme support, allowing users to connect to MUD servers with a single click. This addresses the long-standing feature request in issue #689 from 2013.

## What this PR does

- **Adds telnet:// URI parsing and handling** in command-line arguments
- **Implements intelligent profile matching** - reuses existing profiles when available
- **Auto-creates new profiles** when connecting to unknown servers
- **Provides cross-platform OS integration** for browser and system-wide telnet:// link support
- **Handles multi-instance coordination** - routes URIs to existing Mudlet instances

## How it works

### User Experience
1. User clicks a `telnet://server.com:port` link in browser or application
2. Mudlet automatically opens (or receives the URI if already running)
3. If a matching profile exists → loads the most recently used one
4. If no profile exists → creates new profile with server details and connects
5. User is immediately connected to the MUD server


### Key Features
- **URI Format Support**: `telnet://host[:port]` with port defaulting to 23
- **Profile Matching**: Host + port based matching with existing profiles
- **Smart Naming**: Profiles named as "hostname" or "hostname:port" for non-standard ports
- **Collision Handling**: Automatic numbering for duplicate profile names
- **Multi-instance**: Routes to existing Mudlet instance when available

## Testing

The implementation has been tested with:
- Various telnet:// URI formats (`telnet://server.com`, `telnet://server.com:4000`)
- Profile creation and reuse scenarios
- Multi-instance coordination
- Real MUD server connections (e.g., `telnet://aardmud.org:4000`)

### Test Commands
```bash
# Test basic connection with port
./mudlet "telnet://aardmud.org:4000"

# Test default port (23)
./mudlet "telnet://batmud.bat.org"

# Test profile reuse (run twice)
./mudlet "telnet://testserver.example.com:1234"
```


https://github.com/user-attachments/assets/981cb718-5587-4fcb-a210-c182430b79a4


Players clicking these links will have Mudlet automatically open and connect.

## Closes

Fixes #689

/claim #689
## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Enhancement to existing feature
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

MAVRICK-1 committed:
Add telnet:// URI scheme support for one-click MUD connections

- Implement telnet:// URI parsing and handling in command-line arguments
- Add intelligent profile matching that reuses existing profiles when available
- Auto-create new profiles when connecting to unknown servers
- Provide cross-platform OS integration for browser and system-wide telnet:// links
- Handle multi-instance coordination to route URIs to existing Mudlet instances
- Support standard telnet://host[:port] format with port defaulting to 23
- Add profile naming based on server (e.g. 'server.com:4000' for non-standard ports)
- Include collision handling with automatic numbering for duplicate names

Closes #689

vadi2 commented:
> Mudlet automatically opens (or receives the URI if already running)

How does this work on Windows and macOS? Wouldn't Mudlet need to be registered with the OS as a service that can handle telnet:// links?

MAVRICK-1 commented on src/main.cpp:
```diff
@@ -687,6 +705,11 @@ int main(int argc, char* argv[])
     settings.setValue("MudletPackage/.", "Mudlet Package");
     settings.setValue("MudletPackage/shell/open/command/.", "mudlet %1");
+    
+    // Register telnet:// URI handler
+    settings.setValue("telnet/.", "URL:Telnet Protocol");
+    settings.setValue("telnet/URL Protocol", "");
+    settings.setValue("telnet/shell/open/command/.", QString("\"%1\" \"%2\"").arg(QCoreApplication::applicationFilePath().replace("/", "\\")).arg("%1"));
```
Windows registration happens automatically when Mudlet starts via
  Registry entries.

vadi2 replied:
Can you confirm it works on Windows? I tried and Mudlet did not launch from a telnet:// link.

MAVRICK-1 replied:
@vadi2 i dont have windows now , but i will test it and will give the review , its working fine in mac and ubuntu

MAVRICK-1 commented on mudlet.desktop:
```diff
@@ -3,9 +3,9 @@ Name=Mudlet
 Terminal=false
 Type=Application
-MimeType=application/zip;
+MimeType=application/zip;x-scheme-handler/telnet;
```
MimeType=application/zip;x-scheme-handler/telnet;
  Linux desktop environments handle this via the `.desktop` file's
  `x-scheme-handler/telnet` MIME type.

MAVRICK-1 commented on mudlet.desktop:
```diff
@@ -3,9 +3,9 @@ Name=Mudlet
 Terminal=false
 Type=Application
-MimeType=application/zip;
+MimeType=application/zip;x-scheme-handler/telnet;
```
MimeType=application/zip;x-scheme-handler/telnet;
  Linux desktop environments handle this via the `.desktop` file's
  `x-scheme-handler/telnet` MIME type.

SlySven replied:
Not just GNU/Linux - it should work for the BSDs as well - I hope. :crossed_fingers:

MAVRICK-1 committed:
Add macOS Info.plist for telnet:// URI scheme registration

Implements CFBundleURLTypes configuration to register Mudlet as a
telnet:// URI handler on macOS. This completes cross-platform URI
scheme support alongside existing Windows (Registry) and Linux
(.desktop file) implementations.

- Add Info.plist.in template with telnet scheme registration
- Configure CMakeLists.txt to use Info.plist template for macOS bundles
- Enables one-click MUD connections from browsers on macOS

Fixes missing macOS URI scheme registration for telnet:// protocol.

MAVRICK-1 commented:
I've added the missing macOS support by creating:
  - Info.plist.in template with CFBundleURLTypes configuration
  - CMakeLists.txt integration to generate the bundle's Info.plist

  macOS will now register Mudlet as a `telnet:// handler` when the app bundle
   is installed, just like other native macOS applications.

  The telnet URI handling code `(mudlet::handleTelnetUri())` was already
  implemented - only the OS registration was missing for macOS.

MAVRICK-1 commented:
cc @vadi2 plz check the above comments

SlySven commented on src/main.cpp:
```diff
@@ -438,15 +438,33 @@ int main(int argc, char* argv[])
+                const bool successful = instanceCoordinator->installPackagesRemotely();
+                if (successful) {
+                    return 0;
+                } else {
```
Minor detail - but this is an "else-after-return" - so the else isn't needed - and in fact the whole `if`can be replaced by: `return successful ? 1 : 0` !

SlySven commented on src/Info.plist.in:
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This whole file looks like some sort of input to something to make an `Info.plist` file - I'm not a MacOS expert so how/where is that done?

MAVRICK-1 replied:
Hey @SlySven!  You're absolutely right that it's an input template file.

  ### What is it?
  This is a **CMake template file** that generates the final `Info.plist` for the macOS app bundle during build time. The `.in` suffix is CMake's convention for template files that need
   variable substitution.

  ### How does it work?
  In our `CMakeLists.txt`, we use CMake's `configure_file()` command to process this template:

  ```cmake
  # in CMakeLists.txt
  if(APPLE)
      configure_file(
          "${CMAKE_CURRENT_SOURCE_DIR}/src/Info.plist.in"
          "${CMAKE_CURRENT_BINARY_DIR}/Info.plist"
          @ONLY
      )
      set_target_properties(${PROJECT_NAME} PROPERTIES
          MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_BINARY_DIR}/Info.plist"
      )
  endif()
```
  What does it do?

  The generated Info.plist gets embedded into the macOS app bundle at Mudlet.app/Contents/Info.plist and serves two main purposes:

  1. App metadata - Basic app information like bundle identifier, version, etc.
  2. URL scheme registration - The crucial part for our telnet:// support:
 
```xml
  <key>CFBundleURLTypes</key>
  <array>
      <dict>
          <key>CFBundleURLName</key>
          <string>Telnet URL</string>
          <key>CFBundleURLSchemes</key>
          <array>
              <string>telnet</string>
          </array>
      </dict>
  </array>
```
  Cross-platform context

  This is the macOS equivalent of the Windows registry entries we add in main.cpp:

  - Windows: Registry entries via QSettings
  - macOS: Info.plist URL types registration
  - Linux: .desktop file associations (not implemented yet)

  So when a user clicks a telnet:// link in Safari or any macOS app, the system knows to launch Mudlet and pass the URL as a command-line argument - exactly what our URI handling code
  expects!

SlySven commented on src/mudlet.cpp:
```diff
@@ -3744,6 +3744,103 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+        QString profileHost = readProfileData(profileName, qsl("url"));
+        QString profilePortStr = readProfileData(profileName, qsl("port"));
+        
+        bool ok;
```
I'd suggest initialising it first
```suggestion
        bool ok = false;
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -3744,6 +3744,103 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+        
+        bool ok;
+        int profilePort = profilePortStr.toInt(&ok);
+        if (!ok) {
```
Check the value as well!
```suggestion
        if (!ok || profilePort < 1 || profilePort > USHRT_MAX) {
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -3744,6 +3744,103 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+    // Parse host and optional port
+    int colonIndex = connectionString.indexOf(':');
+    if (colonIndex != -1) {
+        host = connectionString.left(colonIndex);
```
I'd suggest a different name for this variable - given that we do have a `Host` class a lot of the codebase is littered with `pHost` and `host` as pointers and references to instances of that class. Perhaps `hostAddress` or `hosrUrl` would be appropriate? Thoughts @vadi2?

SlySven commented on src/mudlet.cpp:
```diff
@@ -3744,6 +3744,103 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+        }
+        
+        // Check if this profile matches the telnet URI
+        if (profileHost == host && profilePort == port) {
```
You might want/need to make the URL comparison case-insensitive - IIRC the Internet *isn't* case-sensitive for "addresses"...

SlySven commented on src/mudlet.cpp:
```diff
@@ -3744,6 +3744,103 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+        doAutoLogin(matchingProfile);
+    } else {
+        // Create a new profile for this connection
+        QString newProfileName = host;
```
We have some quite restrictive (and case insensitive) requirements for profile names - see in `(bool) dlgConnectionProfiles::validateProfile()`: where you can see that they have to be in:
```cpp
       const QString allowedChars = qsl(". _0123456789-#&aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ");
```
I think you have to ensure the generated name meets this requirement as well!

SlySven reviewed (commented):
I know it seems counter to the purpose of this PR but does every Mudlet  user WANT to force the association of this scheme (`telnet:`) of URLs with Mudlet?

Also, according to https://learn.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85) overall it might not be ideal to automagically open links the user has clicked on in - presumably their web-browser - without  getting some sort of clearance? Normally we are anti-`QMessageBox` because it can mess with the main application "Event loop" but perhaps this is a case where it might be appropriate?

SlySven commented:
Ah, I didn't spot https://github.com/Mudlet/Mudlet/pull/8044#issuecomment-3194313487 - about handling/generating a `Info.plist`...

vadi2 commented:
> I know it seems counter to the purpose of this PR but does every Mudlet user WANT to force the association of this scheme (telnet:) of URLs with Mudlet?

Yes, that is the idea. Such links don't work at all right now. We could also in the future improve it to be more clever and offer a choice, but atm it's not worthwhile (few to no applications actually register the URL handler, so we'd just be bothering by asking).

Kebap commented:
> we'd just be bothering by asking

Agree, Mudlet probably doesn't need to ask for permission at first. Then a user may want to disable the association afterwards. How can this be done? Not from inside Mudlet, I suppose. So the user then changes the settings in their operating system to open telnet: links with another app and not Mudlet again. Fair enough, all is good. However, the next time they start Mudlet again, what happens? Will Mudlet again ignore the setting, not ask the user, but promote Mudlet to use telnet: links? At this point it starts to get annoying.

1. Solution should be for Mudlet to recognize, if another app is already registered for using telnet: url, then step back from claiming it. 
2. Failing that, another solution would be actual configuration inside Mudlet preferences, whether Mudlet should try and register for telnet: url or not. The default being, Mudlet does, but can be deactivated by players not interested in having Mudlet use telnet: links.

vadi2 commented:
That sounds like a good approach @Kebap

MAVRICK-1 commented:
https://github.com/user-attachments/assets/8a836c2b-a356-4807-827b-c8da0e6828d9

@vadi2 @SlySven @Kebap

MAVRICK-1 committed:
Address code review feedback for telnet URI support

- Fix else-after-return patterns in main.cpp
- Initialize bool variables properly in mudlet.cpp
- Add port range validation (1-USHRT_MAX)
- Rename 'host' variable to 'hostAddress' to avoid confusion with Host class
- Implement case-insensitive host comparison
- Add profile name sanitization using allowed character set
- Add user confirmation dialog for security
- Improve Windows registry handler with conflict detection

MAVRICK-1 committed:
Merge branch 'development' into feature/telnet-uri-support

- Resolve include order conflict between climits and CredentialManager
- Fix CodeFactor formatting issue with closing parenthesis
- Integrate with latest credential management changes

vadi2 committed:
Merge branch 'development' into feature/telnet-uri-support

vadi2 commented on src/mudlet.cpp:
```diff
@@ -3744,37 +3745,73 @@ void mudlet::startAutoLogin(const QStringList& cliProfiles)
+        tr("You are about to open a telnet connection to:\n%1\n\nDo you want to proceed?").arg(uri),
+        QMessageBox::Yes | QMessageBox::No,
+        QMessageBox::No
+    );
```
Leris meant ask for permission to overwrite the telnet:// registration if someone else has already registered it, not asking for permission to open a telnet:// link. Those should be seamless I think.

vadi2 commented:
Windows and Linux don't seem to work. I get this on Ubuntu:

<img width="1617" height="1330" alt="image" src="https://github.com/user-attachments/assets/4e7adcd8-7b2c-48bd-8b9e-c892b815e325" />

MAVRICK-1 commented:
@vadi2 i dont know what's the issue , i will test it again. There some more changes , to be done .

Harrison-Teeg committed:
Improve: Allow users to set map background alpha (#7958)

#### Brief overview of PR changes/additions
- Enable users to set an alpha value for the mapper background in:
Preferences > Mapper colors > Background color
- Make transparent/translucent backgrounds actually work (e.g. 0 for a
fully transparent map background).
- Make the 3D mapper use the color set in Preferences > Mapper colors >
Background color (previously it was permanently black no matter what).
- Properly save map background alpha across sessions.

#### Motivation for adding to Mudlet
This unlocks some really neat GUI options as I personally love having a
map overlay on top of the main console.

#### Other info (issues closed, discussion etc)
The 3D mapper is a bit funky since it is based on QOpenGLWidget which
has some limitations:

> [Limitations and Other
Considerations](https://doc.qt.io/qt-6/qopenglwidget.html#limitations-and-other-considerations)

>Putting other widgets underneath and making the QOpenGLWidget
transparent will not lead to the expected results: The widgets
underneath will not be visible. This is because in practice the
QOpenGLWidget is drawn before all other regular, non-OpenGL widgets, and
so see-through type of solutions are not feasible. Other type of
layouts, like having widgets on top of the QOpenGLWidget, will function
as expected.

>When absolutely necessary, this limitation can be overcome by setting
the
[Qt::WA_AlwaysStackOnTop](https://doc.qt.io/qt-6/qt.html#WidgetAttribute-enum)
attribute on the QOpenGLWidget. Be aware however that this breaks
stacking order, for example it will not be possible to have other
widgets on top of the QOpenGLWidget, so it should only be used in
situations where a semi-transparent QOpenGLWidget with other widgets
visible underneath is required.

So I've gone ahead and made it so setting the alpha value of the map
under 255 toggles transparency on AND the 3D map will always stack on
top. The unfortunate side effect is that everything then gets stuck
behind the 3D mapper. Hopefully this is niche enough to not cause
issues. If it is problematic, users are still free to set the alpha back
to 255 and AlwaysStackOnTop will be disabled.

Pardon the light mode!

https://github.com/user-attachments/assets/0c501008-974b-41bd-b1be-ecba288f4c71

To make background transparent as in the video set the alpha to `0` in:
Preferences > Mapper colors > Background color

---------

Co-authored-by: Zooka <136661366+ZookaOnGit@users.noreply.github.com>

SlySven committed:
Fix: partial revert of  change to link-ptbs-to-dblsqd GH Actions workflow (#8062)

#### Brief overview of PR changes/additions
Revert a change I made in #8015 where I changed a `bash` variable name
into a SCREAMING_SNAKE_CASE one - unfortunately I did not (and still
don't) fully understand how it was processed by the `jq` (command line
JSON query tool) and broke the processing of the DBLSQD feed data and
thus stopped the script from working as intended.

#### Motivation for adding to Mudlet
Get Windows PTBs produced again.

#### Other info (issues closed, discussion etc)
This PR should put the affected bit of the script back to exactly how it
was prior to the PR that broke things!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Infrastructure: fix clangtidy workflow to use Qt 6 (#8063)

<!-- Keep the title short & concise so anyone non-technical can
understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
fix clangtidy workflow to use Qt 6
#### Motivation for adding to Mudlet
so we can get clangtidy feedback once again
#### Other info (issues closed, discussion etc)

Co-authored-by: Vadim Peretokin <vadi2@users.noreply.github.com>

vadi2 committed:
Infrastructure: fix clangtidy-analysis to run on ubuntu latest runner (#8064)

<!-- Keep the title short & concise so anyone non-technical can
understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
fix clangtidy-analysis to run on ubuntu latest runner, as ubuntu 20.04
isn't available anymore
#### Motivation for adding to Mudlet
getting clang-tidy working once again
#### Other info (issues closed, discussion etc)

Co-authored-by: Vadim Peretokin <vadi2@users.noreply.github.com>

Kebap committed:
Infrastructure: Verify documentation is available for new features (#8028)

#### Brief overview of PR changes/additions
Check to see if documentation is still needed, before allowing to merge
a PR.
Label to indicate, that a PR needs documentation, must be tagged
manually.

#### Motivation for adding to Mudlet
Avoid situation when the PR gets merged, the docs aren't getting
finished.
So we've got these undocumented Lua functions that no one will use. 

#### Other info (issues closed, discussion etc)
Per Discord discussion with @ZookaOnGit 

You can test this working by adding/removing the label "needs
documentation" right here, right now

SlySven committed:
Improve: move Ambiguous East Asian Width option control from preferences to setConfig() (#8061)

#### Brief overview of PR changes/additions
A replacement for #8020 that actually works - rather than removing the
ability to control the "Ambiguous East Asian Width Characters"
completely it just removes the control from the preferences and, unlike
that failed PR, moves it to something that the Lua API (and thus the
end-user) can examine and control via the `getConfig(...)` and
`setConfig(...)` interface. The setting is called
`ambiguousEAsianWidthCharacters` and the three values it can have are:
* `"narrow"` (with the same effect as the removed checkbox being
"unchecked")
* `"wide"` (same effect as "checked")
* `"auto"` (same effect as "partially-checked")

It also synchronises the `getConfig()` function (which takes no
arguments and returns all the settings it knows about in a table) in the
`Other.lua` external file. There were several settings that had been
added to the internal C core code but which hadn't been added to the
external Lua function that supplements it to handle a "no-arguments"
usage. Whilst those two are now in step this PR does not make any
guarantees about synchronicity with the C core code `setConfig(...)`
function!

#### Motivation for adding to Mudlet
To remove a not often needed control from the preferences to simplify
them.

#### Other info (issues closed, discussion etc)
Whilst correctly updating the "bulk" `getConfig(...)` in the `Other.lua`
file to support this new (there) setting I found a number of others that
had been added to the C++ core `getConfig(...)` function but which had
not been added in that external Lua file so which would be missed if the
user tried to get ALL the settings available.

This replaces/supplants PR #8058 which can be closed as not required.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

mpconley committed:
Fix: Close Profile menu closes wrong profile in detached windows (#8073)

#### Brief overview of PR changes/additions

Fixed the "Close Profile" menu action in detached windows to close the
correct profile by:
- Updated menu connection to use detached window's own close profile
logic instead of main window's method
- Modified `slot_closeCurrentProfile()` to call `closeProfile()`
directly rather than delegating to main window
- Ensures detached window closes automatically when closing the last
remaining profile

#### Motivation for adding to Mudlet

Users reported that clicking "Close Profile" in a detached window was
closing a profile in the main window instead of the profile in the
detached window where the action was triggered. This created a confusing
user experience where the wrong profile would be closed, leaving the
detached window unchanged.

The fix ensures consistent behavior where "Close Profile" always closes
the profile in the window where it was clicked.

#### Other info (issues closed, discussion etc)

Closes #8070

This fix is complementary to PR #7990 which handled toolbar button
states - this addresses the menu action that was missed in that PR.

---

**Before**


https://github.com/user-attachments/assets/9959e325-18fc-4707-a7d8-b0ec900935c4

---

**After**


https://github.com/user-attachments/assets/8b1cae01-bd03-4a6c-9959-d62eb080fa0d

MAVRICK-1 committed:
Add: streamlined module creation feature (#8039)

# Fix: Add streamlined module creation feature

## Overview
Implements a direct "Create Module" workflow in the trigger editor,
eliminating the clunky export-then-import process described in issue
#566.

## Changes Made

### Files Modified:
- **dlgTriggerEditor.h** - Added new action and slot declarations
- **dlgTriggerEditor.cpp** - Implemented "Create Module" button and
functionality
- **dlgPackageExporter.h** - Added preselection and module creation mode
methods
- **dlgPackageExporter.cpp** - Enhanced to support module creation
workflow

### Key Features:
1. **New "Create Module" button** in trigger editor toolbar (📦 icon)
2. **Context-aware preselection** - automatically selects current item
3. **Module creation mode** - configures package exporter for module
workflow
4. **Auto-installation** - modules are installed immediately after
creation
5. **Streamlined UX** - single-click process from selection to installed
module

## Implementation Details

### dlgTriggerEditor Changes:
- Added `mpCreateModuleAction` toolbar button with package-exporter icon
- Implemented `slot_createModule()` that opens package exporter in
module mode
- Pre-selects current item based on editor view type
(Trigger/Alias/Script/etc.)

### dlgPackageExporter Enhancements:
- Added `setModuleCreationMode()` to configure UI for module creation
- Added preselect methods for all item types (Trigger, Timer, Alias,
Script, Action, Key)
- Modified export completion to auto-install as module when in module
creation mode
- Updated window title and prompts for module creation workflow


## Related Issues
Closes #566
/claim #566

## Screenshots
*Note: Screenshots can be added showing the new Create Module button in
the trigger editor toolbar and the enhanced package exporter dialog in
module creation mode.*
<img width="1920" height="1200" alt="Screenshot from 2025-08-16
20-05-59"
src="https://github.com/user-attachments/assets/560a3e29-1a91-4a58-a930-94474701d861"
/>


https://github.com/user-attachments/assets/dbaa7019-c853-4d33-86be-5c39dd4c1fa4

<img width="1920" height="1200" alt="image"
src="https://github.com/user-attachments/assets/631e202c-7d05-4ba9-8c03-7d33e90d7dfd"
/>

---------

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

mpconley committed:
Fix: buffer underflow crash in OSC sequence processing (#8007)

#### Brief overview of PR changes/additions

Fixes a buffer underflow vulnerability in
`TBuffer::translateToPlainText()` when processing OSC (Operating System
Command) escape sequences that appear at the beginning of incoming data.
Adds bounds checking to prevent accessing `localBuffer[SIZE_MAX]` when
`spanEnd` is 0.

#### Motivation for adding to Mudlet

This crash was discovered through AddressSanitizer and could occur when
MUD servers send OSC sequences (like terminal title changes) at the
start of data packets. The vulnerability could cause application crashes
and potential security issues.

#### Other info (issues closed, discussion etc)

- Adds comprehensive Lua test suite (`TBufferOSC_spec.lua`) to prevent
regression
- Tests cover OSC sequences at buffer boundaries, incomplete sequences,
and edge cases
- Fix preserves original functionality while preventing buffer underflow

**Critical crash location:**
```plaintext
8 mudlet TBuffer::translateToPlainText(std::string&, bool) + 14776 (TBuffer.cpp:661) __asan_report_load1 + 64 localBuffer[spanEnd-1] != '\033' // Buffer underflow when spanEnd == 0
```

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

Harrison-Teeg committed:
Fix: Add alpha to setMapBackgroundColor and getMapBackgroundColor (#8071)

#### Brief overview of PR changes/additions
setMapBackgroundColor now takes int r, g, b[, a] with a=255 if not
provided.
getMapBackgroundColor now returns four int values: r, g, b, a

#### Motivation for adding to Mudlet
Brings these two functions into line with the improved transparency
functionality of Mudlet map backgrounds.

#### Other info (issues closed, discussion etc)

[Documentation](https://wiki.mudlet.org/w/Area_51#getMapBackgroundColor.2C_PR_.238071_open)

---------

Co-authored-by: Zooka <136661366+ZookaOnGit@users.noreply.github.com>

vadi2 committed:
Infrastructure: update performance analysis workflow to Qt6 (#8049)

<!-- Keep the title short & concise so anyone non-technical can
understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Update performance analysis workflow to Qt6
#### Motivation for adding to Mudlet
So it runs again and we can tell if Mudlet's display speed has been
affected
#### Other info (issues closed, discussion etc)

Co-authored-by: Vadim Peretokin <vadi2@users.noreply.github.com>

MAVRICK-1 committed:
Merge branch 'development' into feature/telnet-uri-support

MAVRICK-1 committed:
Enhance telnet:// URI support with cross-platform improvements

- Enhance FileOpenHandler to properly handle telnet:// URIs from file open events
- Add DefaultClientUtils for cross-platform telnet handler registration
- Extend MudletInstanceCoordinator with queueUriOrFile() and openUrisLocally() methods
- Improve URI handling architecture for better separation of concerns
- Support telnet:// and mudlet:// URI schemes in FileOpenHandler

Fixes #689

MAVRICK-1 committed:
Fix missing newlines at end of DefaultClientUtils files

MAVRICK-1 commented:
https://github.com/user-attachments/assets/eba8c188-bb8c-4582-a7a9-90969b1ee6bb

@vadi2 can you please tell me how you are testing it ?

vadi2 commented:
It's in the original issue - http://dmud.thebbs.org/lotflink.htm has a bunch of these links. They're meant to be clicked on from the browser.

MAVRICK-1 commented:
https://github.com/user-attachments/assets/1179a92a-7c83-4242-bc70-e91f196872e2

cc @vadi2

vadi2 commented:
Nice, I assume the same should work from the desktop environment's browser?

MAVRICK-1 commented:
> Nice, I assume the same should work from the desktop environment's browser?

I think so

vadi2 commented:
Nice! Does the same work in Windows too?

MAVRICK-1 commented:
Can't confirm about windows, because I use linux. But this new approach should work

vadi2 commented:
Same here, but the bounty requires Windows (80% of players) and macOS support. Virtualbox / docker is an option.

MAVRICK-1 commented:
Yeah , I have a old device will boot that to check it

Kebap commented:
First off, starting `Mudlet.exe "telnet://some.add.ress/"` on Windows, will start Mudlet with a popup asking if I want to connect to some.add.ress and if Yes, it will create a new profile "some.add.ress" and connect there automatically. That is nice.

If I do the same again, a new profile will be created called "some.add.ress (2)" and I can see that in the list of profiles in the connection window. Alright.

When I click this new profile, the server address is indeed "some.add.ress" as expected.

However the profile name there is listed as "some.add.ress 2" without any brackets.
When I click to view an unrelated profile, now the profile was renamed to be without brackets. That's unfortunate.

If I start Mudlet from command line again like before, no new profile "some.add.ress (3)" or alike will be created. Instead, always number 2 will be used for connection. Not sure if this is deliberate or no.

MAVRICK-1 commented:
> First off, starting `Mudlet.exe "telnet://some.add.ress/"` on Windows, will start Mudlet with a popup asking if I want to connect to some.add.ress and if Yes, it will create a new profile "some.add.ress" and connect there automatically. That is nice.
> 
> If I do the same again, a new profile will be created called "some.add.ress (2)" and I can see that in the list of profiles in the connection window. Alright.
> 
> When I click this new profile, the server address is indeed "some.add.ress" as expected.
> 
> However the profile name there is listed as "some.add.ress 2" without any brackets.
> When I click to view an unrelated profile, now the profile was renamed to be without brackets. That's unfortunate.
> 
> If I start Mudlet from command line again like before, no new profile "some.add.ress (3)" or alike will be created. Instead, always number 2 will be used for connection. Not sure if this is deliberate or no.


@vadi2 what should be the pattern?

vadi2 commented:
> If I do the same again

The latest used matching profile should be used in this case!

By a matching profile I mean one that has the same URL and port as what we clicked on. Because a player could have several profiles for the same game, we should use the latest one - we can find the latest one by checking the current/ directory and seeing which is the latest timestamp.

MAVRICK-1 committed:
Fix telnet URI profile handling issues

- Fix profile names with parentheses not displaying correctly in UI
  The setItemName() function now sets both the custom data role and
  the display text to ensure profile names like 'server (2)' are
  shown correctly with brackets

- Reuse most recently used matching profile instead of creating duplicates
  When a telnet URI is opened, check for existing profiles with the same
  host and port, and use the most recently modified one based on the
  'current' directory timestamp

Fixes #689

MAVRICK-1 commented:
https://github.com/user-attachments/assets/135ccd21-6fd2-42c9-83a6-7a17c9cfc806

@vadi2 @Kebap now the pattern is okay ?

MAVRICK-1 commented:
@Kebap just a heads up, did you tested the current commit in windows? Is it working fine now , without duplication?

Kebap commented:
⚠️ The current version seems to introduce unwanted side effects in the "connection" window when viewing profiles there is now the name of each profile written under each profile icon (and also for the previously hidden test profile), where it was not named before:

<img width="813" height="415" alt="grafik" src="https://github.com/user-attachments/assets/9770a360-5c69-4155-9849-ba6d3556f804" />

Kebap commented:
> @Kebap just a heads up, did you tested the current commit in windows? Is it working fine now , without duplication?

I tried `Mudlet-4.19.1-testing-pr8044-6f11e425-windows-64` now.

I type: `Mudlet.exe "telnet://mg.mud.de"` and no Mudlet comes?
I type: `Mudlet.exe "telnet://"` and no Mudlet comes?
I type: `Mudlet.exe ""` and no Mudlet comes?
I type: `Mudlet.exe` and finally Mudlet comes...?!

edit: It seems like, this may be because, I was running at the time also another different DEV-version from a different PR.
This did not stop this unrelated DEV-version from starting, before. Seems like an unnecessary regression to me?! ⚠️ 

Also, how can I test the clicking a telnet:// link in browser etc maybe? ❓

vadi2 commented:
You can find browser links  here: http://dmud.thebbs.org/lotflink.htm

Kebap commented:
> You can find browser links here: http://dmud.thebbs.org/lotflink.htm

That much is clear, but it does not work for me on Win. 
Does it work for you? You last reported it does not:
https://github.com/Mudlet/Mudlet/pull/8044#issuecomment-3198005640

vadi2 commented:
Nothing has changed in that regard. 

As an alternative, if we can't get telnet:// links to work in the browser again (we tried before in 2024), let's use `mudlet://` and `mud://` as an alternative. I have a suspicion that Windows is blocking telnet:// link registration because it is a well-known protocol.

Kebap commented:
I don't follow that suspicion and would like to actually test clicking telnet:// links. 

Not against adding more protocols for good measure. 

@MAVRICK-1 Is it supposed to work in this commit version?

vadi2 committed:
Merge development into feature/telnet-uri-support

ZookaOnGit commented:
Hey, thanks for your contribution but this has been awarded.
